(ND) CAN interfacing code

Burke-Daniel opened the pull request:
This merge request adds the can manager/interface that was developed for comp 5. It has been refactored a bit, and unit tests have been added using the unity unit testing library. To run the unit tests, do the typical:
```shell
mkdir build && cd build
cmake ..
make
```
All tests should pass, and we can expand upon this going forward for other pieces of code that can be unit tested. @Ian-Carroll I can definitely see an opportunity for your PWM code to be unit tested.

colton-smith commented on Subsurface/Node/Common/Peripherals/Inc/can_manager.h:
```diff
@@ -0,0 +1,70 @@
+
+        uint8_t data_as_uint8_t[8] = { 0 };
+        memcpy(data_as_uint8_t, &data, sizeof(T));
+        for (uint8_t i = 0; i < sizeof(T); i++)
+        {
+            m_tx_buffer[i + position] = data_as_uint8_t[i];
+        }
```
Can this loop be replaced with one memcpy?

```memcpy(m_rx_buffer + position, data_as_uint8_t, sizeof(T))```

Better yet, can lines 51-55 be:

```memcpy(m_tx_buffer + position, &data, sizeof(T))? ```

Burke-Daniel replied:
certainly can, good catch

colton-smith commented on Subsurface/Node/Common/Peripherals/Inc/can_manager.h:
```diff
@@ -0,0 +1,70 @@
+        // Clear the txMsg Buffer
+        memset(m_tx_buffer, 0, 8);
+        // Make sure we won't overflow
+        if (sizeof(T) + position > 8) { return false; }
```
Nice !

colton-smith commented on Subsurface/Node/Common/Peripherals/Inc/can_manager.h:
```diff
@@ -0,0 +1,70 @@
+    }
+};
+
+#endif
```
No newline at EOF

colton-smith replied:
Am I blind or is there no corresponding #if for this?

Burke-Daniel replied:
#ifndef NODE_CAN_MANAGER_H at very top, will add comment to make this more obvious

colton-smith commented on Subsurface/Node/Common/Peripherals/Inc/can_interface.h:
```diff
@@ -0,0 +1,59 @@
+    bool receive(uint8_t *rxData);
+};
+
+#endif
```
No newline at EOF

colton-smith commented on Subsurface/Node/Common/Peripherals/Inc/convert_hal_status.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef NODE_CONVERT_HAL_STATUS_H
```
I like the NODE precursor, are gonna do the same for FC?

colton-smith commented on Subsurface/Node/Common/Peripherals/Src/can_interface.cpp:
```diff
@@ -0,0 +1,82 @@
+
+void CANInterface::setTxMailbox(CanTxMailbox mailbox) 
+{
+    m_tx_mailbox = static_cast<uint32_t>(mailbox);
```
Could avoid the cast by having the field be a CanTxMailbox type

colton-smith commented on Subsurface/Node/Common/Peripherals/Src/can_interface.cpp:
```diff
@@ -0,0 +1,82 @@
+
+bool CANInterface::transmit(uint8_t *txData)
+{
+    return halStatusToBool( HAL_CAN_AddTxMessage(m_can_handle, &m_tx_header, txData, &m_tx_mailbox) );
```
Of course, by changing it to a mailbox type, the you'd have to cast it to a uint8_t here so I guess it makes pretty much 0 difference

Burke-Daniel replied:
I like using the defined enums as member variables in the class though, so I've made the change and added the cast here

colton-smith commented on Subsurface/Node/Common/Peripherals/Src/can_interface.cpp:
```diff
@@ -0,0 +1,82 @@
+    switch (fifo)
+    {
+        case CanRxFifo::FIFO0:
+            m_rx_fifo = CAN_RX_FIFO0;
```
should have a break here, I don't see how this will ever allow you to set CAN_RX_FIFO0? Seems like it will always just execute the last case and overwrite

colton-smith commented on Subsurface/Node/Common/Peripherals/Src/can_manager.cpp:
```diff
@@ -0,0 +1,40 @@
+#include "Peripherals/Inc/can_manager.h"
+
+
```
Previously one space was used here

colton-smith commented on Subsurface/Node/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+set(SOURCES
+    Unity/src/Unity.c
+    ../Common/Peripherals/Src/can_manager.cpp
+    mock/mock_can_interface.cpp
```
For the mock directory, what do you think we should either:
1. Having mock/inc and mock/src
2. Defining everything in one header for mock classes

colton-smith commented on Subsurface/Node/Common/Peripherals/Src/convert_hal_status.cpp:
```diff
@@ -0,0 +1,9 @@
+
+bool halStatusToBool(HAL_StatusTypeDef status)
+{
+    if (status == 0x00U) {
+        return true;
+    }
```
wtf is this java type shit

Burke-Daniel replied:
gross

colton-smith reviewed (changes requested):
Really nice man... smells kinda fishy.... I can smell the kraken experience!

I reviewed a bit of style and the logic - NOT the CAN specific stuff like configuration
parameters. 

A couple of small comments, one thing I didn't mention in the comments was 
the tests/test.cpp. Do we really want one test.cpp with every single test for all peripherals?
Seems like we would want tests/{peripheral or drive}/test.cpp

Then we would need a seperate cmake for each though right? Let me know what you think

P.S. The test and vendors folder are lowercase but Common/Peripherals are uppercase :( Maybe we should unify that one way or the other (rip git diff)

LMK if you have any questions!

Burke-Daniel commented on Subsurface/Node/Common/Peripherals/Inc/can_interface.h:
```diff
@@ -0,0 +1,59 @@
+    // Stop CAN peripheral - returns true if successful
+    bool stop(void);
+
+    // Returns number of empty transmit mailboxes
+    uint8_t getNumEmptyTxMailboxes(void);
+    // Set transmit mailbox to CAN_TX_MAILBOX0,1,2
```
I think yes to empty line even with comments

Burke-Daniel commented on Subsurface/Node/Common/Peripherals/Inc/can_manager.h:
```diff
@@ -0,0 +1,70 @@
+    uint8_t* getRxBuffer() { return m_rx_buffer; }
+
+#ifdef UNIT_TEST
+    void setRxBuffer(uint8_t* buffer, int len)
+    {
+        memset(m_rx_buffer, 0, 8);
+        for (int i = 0; i < len; ++i)
+        {
+            m_rx_buffer[i] = buffer[i];
+        }
+    }
```
That's correct, will move definition to source

Burke-Daniel commented on Subsurface/Node/Common/Peripherals/Inc/can_manager.h:
```diff
@@ -0,0 +1,70 @@
+#ifdef UNIT_TEST
+    void setRxBuffer(uint8_t* buffer, int len)
+    {
+        memset(m_rx_buffer, 0, 8);
```
technically no, but in the case that the buffer that is passed in has a length of less than 8, then however m_rx_buffer was set before from len to 8 will be whatever it was before. Just for consistency it is nicer I think to clear the buffer here so that we know it contains all zeroes at the start

Burke-Daniel commented on Subsurface/Node/Common/Peripherals/Inc/can_manager.h:
```diff
@@ -0,0 +1,70 @@
+#endif // UNIT_TEST
+
+    template<typename T>
+    bool txMsgSet(T data, uint8_t position = 0)
```
I believe at work there is no declaration then definition of these, just defined "in place"

Burke-Daniel commented on Subsurface/Node/Common/Peripherals/Inc/can_manager.h:
```diff
@@ -0,0 +1,70 @@
+        // Make sure we won't overflow
+        if (sizeof(T) + position > 8) { return false; }
+
+        uint8_t data_as_uint8_t[8] = { 0 };
```
Yup it does, this is the C style of zero-assigning an array

Burke-Daniel commented:
@colton-smith ready for you to have a look here again, addressed all minor comments and changed directory structure and cmake to allow for an executable for each peripheral/driver being unit tested. Gonna leave unity in here, but going forward will do separate PRs for adding third party libs

colton-smith reviewed (approved):
GTM
